### PR TITLE
Migration to Webpack 5

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Upgraded the version of webpack. See ckeditor/ckeditor5#5815.

### CODE REVIEW

**Comment by pomek** [comment]
Ready for review.

**Comment by pomek** [comment]
Changes in our webpack plugin work also with webpack 4.

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js [hunk 1]
```diff
@@ -39,7 +39,7 @@ module.exports = function getKarmaConfig( options ) {
 		basePath,
 
 		// Frameworks to use. Available frameworks: https://npmjs.org/browse/keyword/karma-adapter
-		frameworks: [ 'mocha', 'sinon-chai' ],
+		frameworks: [ 'mocha', 'sinon-chai', 'webpack' ],
```
Comment (Reinmar) [inline packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js:42]: Why is this change needed? Which library upgrade requires it?

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getwebpackconfig.js [hunk 2]
```diff
@@ -37,14 +37,17 @@ module.exports = function getWebpackConfigForAutomatedTests( options ) {
 								}
 							}
 						},
+						'css-loader',
```
Comment (Reinmar) [inline packages/ckeditor5-dev-tests/lib/utils/automated-tests/getwebpackconfig.js:40]: Which library upgrade requires this?

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getwebpackconfig.js [hunk 3]
```diff
@@ -55,7 +58,7 @@ module.exports = function getWebpackConfigForAutomatedTests( options ) {
 				{
 					test: /\.js$/,
 					loader: require.resolve( '../ck-debug-loader' ),
-					query: {
+					options: {
```
Comment (Reinmar) [inline packages/ckeditor5-dev-tests/lib/utils/automated-tests/getwebpackconfig.js:61]: This is a change in webpack, right?

File: packages/ckeditor5-dev-tests/lib/utils/manual-tests/getwebpackconfig.js [hunk 4]
```diff
@@ -54,7 +54,7 @@ module.exports = function getWebpackConfigForManualTests( options ) {
 		module: {
 			rules: [
 				{
-					test: /\.svg$/,
+					test: /\.(svg|txt|html)$/,
```
Comment (Reinmar) [inline packages/ckeditor5-dev-tests/lib/utils/manual-tests/getwebpackconfig.js:57]: Why is this change needed?

**Review by Reinmar — COMMENTED** [review]
I've left questions about origin of particular changes as this is what's most unclear to me.

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getwebpackconfig.js (hunk 2 above)
Comment (Reinmar) [reply packages/ckeditor5-dev-tests/lib/utils/automated-tests/getwebpackconfig.js:40]: Answering my own question - that's because of postcss-loader v4 as mentioned here: https://github.com/ckeditor/ckeditor5/pull/8213/files#diff-dec446e3fa51a26c473acc17b1283060f6cc8ef8abed2752c70828e88a0bbb5bR219.

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js (hunk 1 above)
Comment (pomek) [reply packages/ckeditor5-dev-tests/lib/utils/automated-tests/getkarmaconfig.js:42]: See: https://github.com/ryanclark/karma-webpack/commit/8e69ed3c4ce3ec12ff2d86612f05566a704a9b52.

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getwebpackconfig.js (hunk 2 above)
Comment (pomek) [reply packages/ckeditor5-dev-tests/lib/utils/automated-tests/getwebpackconfig.js:40]: See: https://github.com/webpack-contrib/postcss-loader/releases/tag/v4.0.0.

> loader output only CSS, so you need to use css-loader/file-loader/raw-loader to inject code inside bundle

File: packages/ckeditor5-dev-tests/lib/utils/automated-tests/getwebpackconfig.js (hunk 3 above)
Comment (pomek) [reply packages/ckeditor5-dev-tests/lib/utils/automated-tests/getwebpackconfig.js:61]: Yes.

File: packages/ckeditor5-dev-tests/lib/utils/manual-tests/getwebpackconfig.js (hunk 4 above)
Comment (pomek) [reply packages/ckeditor5-dev-tests/lib/utils/manual-tests/getwebpackconfig.js]: It isn't. I reverted this in ac10a69.

**Comment by vbartusevicius** [comment]
any news on this?

**Comment by vbartusevicius** [comment]
@pomek - are you going to finish this?

**Comment by netiul** [comment]
Any news on this?

**Comment by marmorag** [comment]
Hi, is there any updates on this? It's blocking us from updating to Webpack 5 and Symfony Webpack Encore 1 :disappointed: 
Thanks!

**Comment by kl3sk** [comment]
Hello, thank you for all of your work.

I agree @marmorag, I'm in the the situation.

**Comment by coolbaluk** [comment]
Would be great if you manage to get this through soon,

**Comment by pomek** [comment]
Hey everyone, thanks a lot for your patience and sorry for the delay in processing this. We have opened a few bigger topics at the same time and with limited resources, we were unable to proceed with the webpack 5 integration. Apologies to everyone waiting for this.

It seems to me that this PR won't be merged because it touched too many layers of our projects in a single PR:

- It brings support for webpack 5 but it also removes support for webpack 4 which is not the right thing to do. We shouldn't force people to upgrade their environments. 
- The PR upgrades our dev tools which don't need to be upgraded. For development purposes, we can still use webpack 4.
- The PR upgrades many libraries to their latest versions which brings additional breaking changes that can be hard to document. Also, the changes are not required for making `@ckeditor/ckeditor5-dev-webpack-plugin` compatible with webpack 5.


Here is a new PR – https://github.com/ckeditor/ckeditor5-dev/pull/692 – that makes the `@ckeditor/ckeditor5-dev-webpack-plugin` package compatible with both versions of webpack (4 and 5).

I hope we will be able to review and release it soon. The changes proposed in this PR should hopefully be sufficient to allow you to use webpack 5 and CKEditor 5. Once again, apologies for any inconvenience caused!

**Comment by kl3sk** [comment]
Thanks for your feedback @pomek, I think we dont wait for __this PR__ but for a __Webpack5__ compatiblity 😉.

Hope the new PR come (very) soon.

**Comment by pomek** [comment]
> a Webpack5 compatiblity

Good news! Please, see https://github.com/ckeditor/ckeditor5-dev/releases/tag/v24.3.0. 
There is a new release of the [`@ckeditor/ckeditor5-dev-webpack-plugin`](https://www.npmjs.com/package/@ckeditor/ckeditor5-dev-webpack-plugin) package that supports webpack 5 (and of course webpack 4).

**Comment by pomek** [comment]
Closing due to https://github.com/ckeditor/ckeditor5-dev/pull/673#issuecomment-786715895.
